### PR TITLE
Bfd test multihop

### PR DESCRIPTION
**Pull request by siqbal1986**
### Description of PR
BFD multihop  test.

Summary:
The openBFDD being used to verify BFD functionality in sonic does not support multi-hop BFD sessions.
This PR adds a script which acts as a responder and initiator of BFD multi-hop sessions. This script is used by a new bfd multi-hop test added in the existing test suite.
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

#### What is the motivation for this PR?
Test multihop BFD session capabilities of Cisco and other devices.
#### How did you do it?
Added a Script which sends and intercepts BFD packets.
#### How did you verify/test it?
Manual verification using a DUT.

### CODE REVIEW

**Comment by prsunny** [comment]
As discussed, please fix the test to have routed IP as remote IP instead of P2P interface IP so as to simulate true multhop sessions

File: ansible/roles/test/files/helpers/bfd_responder.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+from pprint import pprint
+import logging
+logging.getLogger("scapy.runtime").setLevel(logging.ERROR)
+import ptf.packet as scapy
+import scapy.all as scapy2
+scapy2.conf.use_pcap=True
+import scapy.arch.pcapdnet
+from  scapy.contrib.bfd import BFD
+def hexdump(data):
+    print " ".join("%02x" % ord(d) for d in data)
```
Comment (wangxin) [inline ansible/roles/test/files/helpers/bfd_responder.py:18]: Can you make this print compatible with python3?

Comment (siqbal1986) [reply ansible/roles/test/files/helpers/bfd_responder.py:18]: done

File: tests/bfd/test_bfd.py [hunk 2]
```diff
@@ -185,78 +311,118 @@ def remove_bfd_sessions(duthost, local_addrs, neighbor_addrs):
...
+                check_dut_bfd_status(duthost, neighbor_addr, "Up")
+                check_ptf_bfd_status(ptfhost, neighbor_addr, local_addrs[idx], "Up")
+
+        update_bfd_session_state(ptfhost, neighbor_addrs[update_idx], local_addrs[update_idx], "up")
+        time.sleep(1)
+
+        check_dut_bfd_status(duthost, neighbor_addrs[update_idx], "Up")
+        check_ptf_bfd_status(ptfhost, neighbor_addrs[update_idx], local_addrs[update_idx], "Up")
+
+
```
[the remark this answers is not in the document]
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:353]: fixed

File: tests/bfd/test_bfd.py [hunk 3]
```diff
@@ -185,78 +311,118 @@ def remove_bfd_sessions(duthost, local_addrs, neighbor_addrs):
...
+        del_ipaddr(ptfhost, neighbor_addrs, prefix_len, neighbor_interfaces, ipv6)
+        remove_bfd_sessions(duthost, neighbor_addrs)
+        remove_dut_ip(duthost, neighbor_devs, local_addrs, prefix_len)
+
+@pytest.mark.parametrize('ipv6', [False, True], ids=['ipv4', 'ipv6'])
+def test_bfd_multihop(rand_selected_dut, ptfhost, tbinfo, toggle_all_simulator_ports_to_rand_selected_tor_m, ipv6):
+    duthost = rand_selected_dut
+
+    bfd_session_cnt = 5
+    skip_201911_and_older(duthost)
```
[the remark this answers is not in the document]
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:406]: Can you please elaborate this a bit. I do not understand how it should be done.

File: tests/bfd/test_bfd.py [hunk 4]
```diff
@@ -201,15 +241,57 @@ def create_bfd_sessions(ptfhost, duthost, local_addrs, neighbor_addrs, dut_init_
     if dut_init_first :
         ptfhost.shell(ptf_buffer)
 
+def create_bfd_sessions_multihop(ptfhost, duthost, loopback_addr, ptf_intf, nexthop_ip, neighbor_addrs):
```
Comment (ihorchekh) [inline tests/bfd/test_bfd.py:244]: `nexthop_ip` is not used in this function.

File: tests/bfd/test_bfd.py [hunk 5]
```diff
@@ -312,5 +391,34 @@ def test_bfd_scale(rand_selected_dut, ptfhost, tbinfo, toggle_all_simulator_port
...
+    try:
+        cmd_buffer = ""
+        for neighbor in neighbor_addrs:
+            cmd_buffer += 'sudo ip route add {} via {} ;'.format(neighbor, nexthop_ip)
+        duthost.shell(cmd_buffer)
+
+        create_bfd_sessions_multihop(ptfhost, duthost, loopback_addr, ptf_intf, nexthop_ip, neighbor_addrs)
+
+        time.sleep(1)
+        for idx, neighbor_addr in enumerate(neighbor_addrs):
```
Comment (ihorchekh) [inline tests/bfd/test_bfd.py:413]:
```suggestion
        for neighbor_addr in neighbor_addrs:
```

File: tests/bfd/test_bfd.py [hunk 6]
```diff
@@ -312,5 +391,34 @@ def test_bfd_scale(rand_selected_dut, ptfhost, tbinfo, toggle_all_simulator_port
...
-        remove_bfd_sessions(duthost, local_addrs, neighbor_addrs)
+        remove_bfd_sessions(duthost, neighbor_addrs)
         remove_dut_ip(duthost, neighbor_devs, local_addrs, prefix_len)
+
+@pytest.mark.parametrize('ipv6', [False, True], ids=['ipv4', 'ipv6'])
+def test_bfd_multihop(request, rand_selected_dut, ptfhost, tbinfo, toggle_all_simulator_ports_to_rand_selected_tor_m, ipv6):
+    duthost = rand_selected_dut
+
+    bfd_session_cnt = int(request.config.getoption('--num_sessions'))
+    skip_201911_and_older(duthost)
```
Comment (ihorchekh) [inline tests/bfd/test_bfd.py:402]:
```suggestion
```

File: tests/bfd/test_bfd.py [hunk 7]
```diff
@@ -312,5 +391,34 @@ def test_bfd_scale(rand_selected_dut, ptfhost, tbinfo, toggle_all_simulator_port
...
+    loopback_addr, ptf_intf, nexthop_ip, neighbor_addrs = get_neighbors_multihop(duthost, tbinfo, ipv6, count = bfd_session_cnt)
+    try:
+        cmd_buffer = ""
+        for neighbor in neighbor_addrs:
+            cmd_buffer += 'sudo ip route add {} via {} ;'.format(neighbor, nexthop_ip)
+        duthost.shell(cmd_buffer)
+
+        create_bfd_sessions_multihop(ptfhost, duthost, loopback_addr, ptf_intf, nexthop_ip, neighbor_addrs)
```
Comment (ihorchekh) [inline tests/bfd/test_bfd.py:410]:
```suggestion
        create_bfd_sessions_multihop(ptfhost, duthost, loopback_addr, ptf_intf, neighbor_addrs)
```

File: tests/bfd/test_bfd.py [hunk 8]
```diff
@@ -12,6 +12,9 @@
 pytestmark = [
     pytest.mark.topology('t1', 't1-lag', 't1-64-lag')
 ]
+BFD_RESPONDER_SCRIPT_SRC_PATH = '../ansible/roles/test/files/helpers/bfd_responder.py'
+BFD_RESPONDER_SCRIPT_DEST_PATH = '/opt/bfd_responder.py'
+
 def skip_201911_and_older(duthost):
```
Comment (ihorchekh) [inline tests/bfd/test_bfd.py:18]:
```suggestion
```
This function can be removed, because of this PR #5708

**Review by ihorchekh — DISMISSED** [review]

File: tests/bfd/test_bfd.py (hunk 4 above)
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:250]: done

File: tests/bfd/test_bfd.py (hunk 5 above)
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:422]: done

File: tests/bfd/test_bfd.py (hunk 8 above)
Comment (siqbal1986) [reply tests/bfd/test_bfd.py]: same as above

File: tests/bfd/test_bfd.py (hunk 7 above)
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:419]: already done

File: tests/bfd/test_bfd.py (hunk 6 above)
Comment (siqbal1986) [reply tests/bfd/test_bfd.py:411]: already fixed.

File: tests/bfd/test_bfd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Blueve) [reply tests/bfd/test_bfd.py:411]: @siqbal1986 This is the doc: https://github.com/sonic-net/sonic-mgmt/tree/master/tests/common/plugins/conditional_mark

and this is a sample:
https://github.com/sonic-net/sonic-mgmt/blob/7185455426370fe7ee5318f7c6772bcc873bcee7/tests/common/plugins/conditional_mark/tests_mark_conditions.yaml#L420

**Comment by siqbal1986** [comment]
@ihorchekh if you are ok with this.Please approve the PR.

**Comment by siqbal1986** [comment]
@wangxin Can you please approve it so I can get it merged?

File: tests/bfd/test_bfd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (siqbal1986) [reply tests/bfd/test_bfd.py]: i have fixed this by removing the statement. There was already a conditional mark for this test in the tests_mark_conditions.yaml

**Comment by wangxin** [comment]
@siqbal1986 I noticed that LGTM raised 3 alerts. Can you address them?

File: ansible/roles/test/files/helpers/bfd_responder.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+import argparse
+import os.path
+from fcntl import ioctl
+import logging
+logging.getLogger("scapy.runtime").setLevel(logging.ERROR)
+import scapy.all as scapy2
+scapy2.conf.use_pcap=True
+from  scapy.contrib.bfd import BFD
+def hexdump(data):
+    print (" ".join("%02x" % ord(d) for d in data))
```
Comment (wangxin) [inline ansible/roles/test/files/helpers/bfd_responder.py:15]: This print is not python3 compatible.

File: ansible/roles/test/files/helpers/bfd_responder.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+        if bfd_state== 3L:
+            interface.send(session["pkt"])
+            return
+
+        if bfd_state == 2L:
+            return
+        session = self.sessions[ip_dst]
+        session["other_disc"] = bfd_remote_disc
+        bfd_pkt_init = self.craft_bfd_packet( session, data, mac_src, mac_dst, ip_src, ip_dst, bfd_remote_disc, 2L )
+        print "sending INIT", bfd_pkt_init
```
Comment (wangxin) [inline ansible/roles/test/files/helpers/bfd_responder.py:99]: This print is incompatible with python3.

File: ansible/roles/test/files/helpers/bfd_responder.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    parser.add_argument('--conf', '-c', type=str, dest='conf', default='/tmp/from_t1.json', help='path to json file with configuration')
+    args = parser.parse_args()
+
+    return args
+
+def main():
+    args = parse_args()
+
+    if not os.path.exists(args.conf):
+        print "Can't find file %s" % args.conf
```
Comment (wangxin) [inline ansible/roles/test/files/helpers/bfd_responder.py:142]: Incompatible with python3.

File: tests/bfd/test_bfd.py [hunk 12]
```diff
@@ -7,6 +7,13 @@
     pytest.mark.topology('t1', 't1-lag', 't1-64-lag')
```
Comment (wangxin) [inline tests/bfd/test_bfd.py:7]: Just mark topology "t1" is enough. Here the mark is about topology type, not topology name.
Both "t1", "t1-lag" and "t1-64-lag" belong to topology type "t1".

Comment (siqbal1986) [reply tests/bfd/test_bfd.py]: fixed

File: ansible/roles/test/files/helpers/bfd_responder.py (hunk 11 above)
Comment (siqbal1986) [reply ansible/roles/test/files/helpers/bfd_responder.py]: fixed

File: ansible/roles/test/files/helpers/bfd_responder.py (hunk 10 above)
Comment (siqbal1986) [reply ansible/roles/test/files/helpers/bfd_responder.py]: fixed

File: ansible/roles/test/files/helpers/bfd_responder.py (hunk 9 above)
Comment (siqbal1986) [reply ansible/roles/test/files/helpers/bfd_responder.py]: removed as this code wasn't running.

**Review by wangxin — APPROVED** [review]

**Pull request merged**